### PR TITLE
accountant: separate modify_balance into separate module and check er…

### PR DESCRIPTION
**Pull request by conorpp**
* Updates tests in accountant to check for specific errors instead of just that the error occurred.  There were a couple of cases the we weren't getting the error we thought we were, but there no actual bugs.

* Add `accountant` governance message type to rust sdk.

* Changes `ModifyBalance` to be checked only when the "global account" module is used in the governance message.  Otherwise, ModifyBalance can be confused with other defined token bridge governance messages.

* Change guardian to use "global accountant" module for `ModifyBalance`.

* Update wormchain tests since they include accountant.

### CODE REVIEW

File: cosmwasm/contracts/global-accountant/src/contract.rs [hunk 1]
```diff
@@ -334,9 +334,21 @@ fn handle_vaa(
...
-        handle_governance_vaa(deps.branch(), info, body.with_payload(govpacket))?
+        let mut module = [0u8; 32];
+        module.clone_from_slice(&body.payload.get()[..32]);
+
+        if let Ok(govpacket) = serde_wormhole::from_slice::<token::GovernancePacket>(body.payload) {
+            handle_token_governance_vaa(deps.branch(), body.with_payload(govpacket))?
+        } else if let Ok(govpacket) =
+            serde_wormhole::from_slice::<accountant_module::GovernancePacket>(body.payload)
+        {
+            handle_accountant_moduleernance_vaa(deps.branch(), info, body.with_payload(govpacket))?
```
Comment (evan-gray) [inline cosmwasm/contracts/global-accountant/src/contract.rs:345]: Typo?

**Review by evan-gray — DISMISSED** [review]

File: sdk/rust/core/src/accountant.rs [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+                                    token_chain,
+                                    token_address,
+                                    kind,
+                                    amount,
+                                    reason,
+                                }
+                            }
+                            v => {
+                                return Err(Error::custom(format_args!(
+                                    "invalid action: {v}, expected one of: 1, 2"
```
Comment (hendrikhofstadt) [inline sdk/rust/core/src/accountant.rs:317]: only expecting 1

File: sdk/rust/core/src/accountant.rs [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+                        token_chain,
+                        token_address,
+                        kind,
+                        amount,
+                        reason,
+                    }
+                }
+                v => {
+                    return Err(Error::custom(format_args!(
+                        "invalid value {v}, expected one of 1, 2"
```
Comment (hendrikhofstadt) [inline sdk/rust/core/src/accountant.rs:236]: of 1

File: sdk/rust/core/src/accountant.rs [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+            payload: GovernancePacket {
+                chain: Chain::Solana,
+                action: Action::ModifyBalance {
+                    sequence: 1,
+                    chain_id: 2,
+                    token_chain: 3,
+                    token_address: Address([0x32u8; 32]),
+                    kind: ModificationKind::Add,
+                    amount: Amount([0x31u8; 32]),
+                    reason: "".into(),
```
Comment (hendrikhofstadt) [inline sdk/rust/core/src/accountant.rs:413]: maybe add a reason to the test

File: cosmwasm/contracts/global-accountant/tests/submit_vaas.rs [hunk 5]
```diff
@@ -181,14 +199,19 @@ fn bad_signature() {
...
+    assert_eq!(
+        "generic error: querier contract error: failed to recover verifying key",
+        err.root_cause().to_string().to_lowercase()
+    );
 }
 
 #[test]
 fn non_transfer_message() {
     let (wh, mut contract) = proper_instantiate();
+    register_emitters(&wh, &mut contract, 4);
```
Comment (hendrikhofstadt) [inline cosmwasm/contracts/global-accountant/tests/submit_vaas.rs:214]: why did this work without the registration / why did you add this?

File: cosmwasm/contracts/global-accountant/tests/chain_registration.rs [hunk 6]
```diff
@@ -88,22 +88,30 @@ fn wrong_target() {
...
     body.emitter_chain = Chain::Fantom;
 
     let (_, data) = sign_vaa_body(&wh, body);
-    contract
+    let err = contract
         .submit_vaas(vec![data])
         .expect_err("successfully executed chain registration with non-governance chain");
+
+    // This tries to parse as a normal token bridge vaa and fails.  Not sure how best to differentiate from a governance message.
+    _ = err;
```
Comment (hendrikhofstadt) [inline cosmwasm/contracts/global-accountant/tests/chain_registration.rs:114]: Why are you explicitly dropping the error and not asserting?

File: cosmwasm/contracts/global-accountant/tests/chain_registration.rs [hunk 7]
```diff
@@ -114,9 +122,12 @@ fn non_governance_emitter() {
     body.emitter_address = Address([0x88; 32]);
 
     let (_, data) = sign_vaa_body(&wh, body);
-    contract
+    let err = contract
         .submit_vaas(vec![data])
         .expect_err("successfully executed chain registration with non-governance emitter");
+
+    // This tries to parse as a normal token bridge vaa and fails.  Not sure how best to differentiate from a governance message.
+    _ = err;
```
Comment (hendrikhofstadt) [inline cosmwasm/contracts/global-accountant/tests/chain_registration.rs:130]: same here?

File: cosmwasm/contracts/global-accountant/tests/chain_registration.rs [hunk 8]
```diff
@@ -88,22 +88,30 @@ fn wrong_target() {
...
+        err.root_cause().to_string().to_lowercase()
+    );
 }
 
 #[test]
 fn non_governance_chain() {
     let (wh, mut contract) = proper_instantiate();
 
     let mut body = create_vaa_body();
+    body.payload.chain = Chain::Solana;
```
Comment (hendrikhofstadt) [inline cosmwasm/contracts/global-accountant/tests/chain_registration.rs:105]: why is this required? This tests the emitter chain check.

**Review by hendrikhofstadt — CHANGES_REQUESTED** [review]
See inline comments

**Review by bruce-riley — DISMISSED** [review]
I only reviewed the guardian changes.

File: node/cmd/guardiand/adminserver.go [hunk 9]
```diff
@@ -148,7 +148,7 @@ func tokenBridgeRegisterChain(req *nodev1.BridgeRegisterChain, timestamp time.Ti
 
 // tokenBridgeModifyBalance converts a nodev1.TokenBridgeModifyBalance message to its canonical VAA representation.
 // Returns an error if the data is invalid.
-func tokenBridgeModifyBalance(req *nodev1.BridgeModifyBalance, timestamp time.Time, guardianSetIndex uint32, nonce uint32, sequence uint64) (*vaa.VAA, error) {
+func accountantModifyBalance(req *nodev1.AccountantModifyBalance, timestamp time.Time, guardianSetIndex uint32, nonce uint32, sequence uint64) (*vaa.VAA, error) {
```
Comment (panoel) [inline node/cmd/guardiand/adminserver.go:151]: Please change comment on line 149 to reflect code changes.

File: cosmwasm/contracts/global-accountant/src/contract.rs [hunk 10]
```diff
@@ -334,9 +334,21 @@ fn handle_vaa(
     let mut evt = if body.emitter_chain == Chain::Solana
         && body.emitter_address == wormhole::GOVERNANCE_EMITTER
     {
-        let govpacket = serde_wormhole::from_slice(body.payload)
-            .context("failed to parse governance packet")?;
-        handle_governance_vaa(deps.branch(), info, body.with_payload(govpacket))?
+        let mut module = [0u8; 32];
+        module.clone_from_slice(&body.payload.get()[..32]);
+
+        if let Ok(govpacket) = serde_wormhole::from_slice::<token::GovernancePacket>(body.payload) {
```
Comment (a5-pickle) [inline cosmwasm/contracts/global-accountant/src/contract.rs:340]: This might be better written as `match` where the default (`_`) match results in `bail!`.

`if let` is a nice syntax when there aren't other conditions (either match something specific, otherwise do something else).

File: cosmwasm/contracts/global-accountant/src/contract.rs [hunk 11]
```diff
@@ -378,7 +389,22 @@ fn handle_governance_vaa(
...
+}
+
+fn handle_accountant_moduleernance_vaa(
+    deps: DepsMut<WormholeQuery>,
+    info: &MessageInfo,
+    body: Body<accountant_module::GovernancePacket>,
+) -> anyhow::Result<Event> {
+    ensure!(
+        body.payload.chain == Chain::Any || body.payload.chain == Chain::Wormchain,
+        "this governance VAA is for another chain"
```
Comment (a5-pickle) [inline cosmwasm/contracts/global-accountant/src/contract.rs:403]: It would be nice if this error message were distinct from `handle_token_governance_vaa` when it checks its governance target chain.

File: sdk/rust/core/src/token.rs [hunk 12]
```diff
@@ -469,28 +369,6 @@ mod governance_packet_impl {
...
-                        token_chain,
-                        token_address,
-                        kind,
-                        amount,
-                        reason,
-                    }
-                }
                 v => {
                     return Err(Error::custom(format_args!(
                         "invalid value: {v}, expected one of 1, 2, 3"
```
Comment (a5-pickle) [inline sdk/rust/core/src/token.rs:374]: This needs to change to either "of 1, 2" or the error message should just be `"invalid value: {v}"`.

If the latter is chosen, we can change the same error in _accountant.rs_.

**Review by a5-pickle — CHANGES_REQUESTED** [review]
Reviewed cosmwasm related changes and left comments.

File: cosmwasm/contracts/global-accountant/src/contract.rs (hunk 10 above)
Comment (conorpp) [reply cosmwasm/contracts/global-accountant/src/contract.rs:340]: match won't work here since we need to actually call `serde_wormhole::from_slice:` twice.  Once for `token::GovernancePacket` and once for `accountant::GovernancePacket`.

File: cosmwasm/contracts/global-accountant/tests/chain_registration.rs (hunk 8 above)
Comment (conorpp) [reply cosmwasm/contracts/global-accountant/tests/chain_registration.rs:105]: I think I added this debugging, just removed.

**Review by a5-pickle — DISMISSED** [review]
Out of scope for this PR, but we may want to move more deserialized output checking to the Rust SDK as opposed to the contract.

Because the Rust SDK already checks for the module name in its deserialization of governance messages, we may want to enforce target chain checking in this deserialization, too. Otherwise we should have all the deserialization output checked in the contract.

File: cosmwasm/contracts/global-accountant/src/contract.rs [hunk 13]
```diff
@@ -334,9 +334,15 @@ fn handle_vaa(
     let mut evt = if body.emitter_chain == Chain::Solana
         && body.emitter_address == wormhole::GOVERNANCE_EMITTER
     {
```
Comment (jynnantonix) [inline cosmwasm/contracts/global-accountant/src/contract.rs:336]: If we want to still handle the errors from parsing, we could do something like:

```rust
    if body.payload.len() < 32 {
        bail!("governance module missing");
    }
    let module = &body.payload[..32];

    if module == TOKENBRIDGE_MODULE {
        let govpacket = serde_wormhole::from_slice(body.payload)
            .context("failed to parse tokenbridge governance packet")?;
        handle_token_governance_vaa(deps.branch(), body.with_payload(govpacket))
    } else if module == ACCOUNTANT_MODULE {
        let govpacket = serde_wormhole::from_slice(body.payload)
            .context("failed to parse accountant governance packet")?;
        handle_accountant_governance_vaa(deps.branch(), body.with_payload(govpacket))
    } else {
        bail!("unknown governance module")
    }
```

We could export the constants for the tokenbridge and accountant governance modules from the rust SDK.

File: sdk/rust/core/src/accountant.rs [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+        where
+            D: Deserializer<'de>,
+        {
+            let arr = <[u8; 32]>::deserialize(deserializer)?;
+
+            if arr == MODULE {
+                Ok(Module)
+            } else {
+                Err(Error::custom(
+                    "invalid governance module, expected \"NFTBridge\"",
```
Comment (jynnantonix) [inline sdk/rust/core/src/accountant.rs:126]: nit: "NFTBridge" -> "GlobalAccountant"

**Review by jynnantonix — APPROVED** [review]

**Comment by jynnantonix** [comment]
> Out of scope for this PR, but we may want to move more deserialized output checking to the Rust SDK as opposed to the contract.
> 
> Because the Rust SDK already checks for the module name in its deserialization of governance messages, we may want to enforce target chain checking in this deserialization, too. Otherwise we should have all the deserialization output checked in the contract.

I think there's a difference in the 2 types of checks.  The module name of governance messages is checked during deserialization for type-safety reasons.  Tokenbridge governance messages **always** have the tokenbridge governance module prefixed in the serialized payload and anything that has a different prefix is a completely different type and so should not be parsed as a tokenbridge governance action.  This is always true regardless of the context in which the message is deserialized.

On the other hand, most of the other checks depend on the context: if it's a contract on a specific chain we want to check that the target chain matches or if it's off-chain then we may not care about the exact value of the target chain at all (only that it is one of the supported chains).  These are application-safety checks, which is why they are not done during deserialization but afterwards.

Or to put it another way, the wormhole wire format is generally not self-describing except in the case of governance messages, which is why we take advantage of that to provide additional type safety that's not normally possible for other kinds of messages.

So while I'm not opposed to moving more checks into the SDK we should be careful to not mix up application-safety checks with type-safety checks.

**Review by hendrikhofstadt — APPROVED** [review]

**Review by bruce-riley — DISMISSED** [review]

File: cosmwasm/contracts/global-accountant/src/contract.rs (hunk 13 above)
Comment (conorpp) [reply cosmwasm/contracts/global-accountant/src/contract.rs:336]: extra safe, will add it now.

**Review by hendrikhofstadt — APPROVED** [review]

**Review by jynnantonix — APPROVED** [review]

**Review by bruce-riley — APPROVED** [review]
